### DISCUSSION TITLE
Consult-buffer integration

### DISCUSSION THREAD

**Discussion by axelknock** [alphapapa/activities.el#59] [Ideas]
`activities-switch-buffer` works well, but I could not find a way to integrate it with [consult](https://github.com/minad/consult)'s `consult-buffer` in a way where only buffers from the current activity are shown.

That being said, if it's not within the design scope, the package [bufferlo](https://github.com/florommel/bufferlo) provides provides for showing only tab (and frame) local buffers, which works well with `activities-tabs-mode`, and provides an integration with consult.

I might be missing something obvious, but thought I would post an issue about it in case anyone looks for a similar solution.

**Comment by alphapapa** [comment]
> `activities-switch-buffer` works well, but I could not find a way to integrate it with [consult](https://github.com/minad/consult)'s `consult-buffer` in a way where only buffers from the current activity are shown.

What have you tried?  The relevant code is here: https://github.com/alphapapa/activities.el/blob/65bfb6f7f246a66a7b6209096c3a634430323758/activities-tabs.el#L100-L110

Here's what I'm using now for `C-x b`:

```el
(defun ap/switch-to-buffer ()
    "Switch to buffer offered from various sources.
If an activity is current, use `activities-switch-buffer';
otherwise use `bufler-switch-buffer'."
    (interactive)
    (cond ((or (equal '(16) current-prefix-arg)
               (not (activities-current)))
           (call-interactively #'bufler-switch-buffer))
          (t (call-interactively #'activities-switch-buffer))))
```
I also use `marginalia` and `vertico`, which gives me a UI like this:

![example](https://github.com/alphapapa/activities.el/assets/601365/54650a12-1c3f-4a89-8b13-b06ecbce7d58)

**Comment by axelknock** [comment]
Consult's `consult-buffer` consolidates various sources of buffers into one completing-read. Inspired by Bufferlo, I was able to use some of the existing internal functions of activities.el to write a function `activities-local-buffer-p`, which returns non-nil when the buffer is associated with the activity returned by `activities-current`. Consult has a utility function for such a predicate function.

I also added that function as a null predicate for the "all buffers" source, such that they will only show by default outside of activities tabs.
```emacs-lisp
(defun activities-local-buffer-p (buffer)
    "Returns non-nil if BUFFER is present in `activities-current'."
    (when (activities-current)
      (memq buffer (activities-tabs--tab-parameter 'activities-buffer-list (activities-tabs--tab (activities-current))))))

  (defvar my-consult--source-activities-buffer
    `(:name "Activities Buffers"
            :narrow   ?a
            :category buffer
            :face     consult-buffer
            :history  buffer-name-history
            :state    ,#'consult--buffer-state
            :default  t
            :items ,(lambda () (consult--buffer-query
                                :predicate #'activities-local-buffer-p
                                :sort 'visibility
                                :as #'buffer-name)))
    "Activities local buffers candidate source for `consult-buffer'.")

  (defvar my-consult--source-all-buffers
    `(:name "All Buffers"
            :narrow   ?b
            :category buffer
            :face     consult-buffer
            :history  buffer-name-history
            :state    ,#'consult--buffer-state
            :enabled ,(lambda () (not (activities-current)))
            :items ,(lambda () (consult--buffer-query :sort 'visibility
                                                      :as #'consult--buffer-pair)))
    "All buffers candidate source for `consult-buffer'.")
  (delete 'consult--source-buffer consult-buffer-sources)
  (add-to-list 'consult-buffer-sources 'my-consult--source-all-buffers)
  (add-to-list 'consult-buffer-sources 'my-consult--source-activities-buffer)
```

Here is a screenshot of the result, also using `vertico` and `marginalia`:
![Screenshot_20240329_220403](https://github.com/alphapapa/activities.el/assets/72520178/a91411d1-5191-434f-82c3-ece45604a5b2)
EDIT: updated screenshot

**Comment by lispcat** [comment]
I found one small issue with the consult integration.

When using `consult-buffer`, for each buffer preview, its corresponding buffer is automatically added to the current activity, as long as it's opened (excludes non-opened buffers, like recent files). So in other words, if I were to narrow the source to include all buffers using `b SPC`, then hover over every buffer, having each buffer preview appear, then every single opened buffer will be added to the current activity.
 
I wonder if this could be fixed by slightly modifying how buffers are registered into the current activity, such that buffers viewed through the `consult-buffer` buffer previews are not added to the current activity...

Edit 1: I'm assuming that for each buffer preview, `activities-mode` is automatically enabled, causing `#'activities-save-all` to run, as per the minor mode definition.

Edit 2: It seems as though for each buffer preview, that buffer is automatically added to the frame's buffers. So one solution could be preventing consult from adding previewing buffers to the frame's buffers...

Edit 3: Found a decent working solution, though it could be made simpler and not depend on `activities-tabs-mode` with more functions from `activities.el`:
```elisp
(with-eval-after-load 'consult
  (defun +consult-buffer--frame-buffers-around (orig-fn &rest args)
    (let ((_log nil)
          (result nil))
      (if-let* ((_predicates (and activities-mode
                                  activities-tabs-mode
                                  tab-bar-mode
                                  (activities-current)))
                (current-tab (tab-bar--current-tab))
                (tab-index (tab-bar--current-tab-index))
                (init-tab-buffers (alist-get 'activities-buffer-list current-tab)))
          ;; if-let then form
          (unwind-protect
              (setq result (apply orig-fn args))
            ;; re-fetch current tab after consult-buffer
            (let* ((updated-tab (tab-bar--current-tab))
                   (post-tab-buffers
                    (seq-filter #'buffer-live-p
                                (alist-get 'activities-buffer-list updated-tab)))
                   (current-buffer (current-buffer))
                   ;; remove new temp buffers from post-tab-buffers
                   (filtered-new
                    (seq-filter (lambda (buf)
                                  (and buf
                                       (or (member buf init-tab-buffers)
                                           (eq buf current-buffer)
                                           (equal (buffer-name buf) " *Minibuf-1*"))))
                                post-tab-buffers)))
              ;; debugging
              (when _log
                (message "LOG: activities: ignore consult-buffer previews: %S"
                         (mapcar #'buffer-name
                                 (seq-remove (lambda (x) (member x filtered-new))
                                             post-tab-buffers))))
              ;; update the tab's activities-buffer-list
              (let ((tabs (funcall tab-bar-tabs-function)))
                (setf (alist-get 'activities-buffer-list (nth tab-index tabs))
                      filtered-new))

              result))
        ;; if-let else form
        (apply orig-fn args))))

  (advice-add #'consult-buffer :around #'+consult-buffer--frame-buffers-around))
```
